doc/config-generic-update-rollback/Json_Patch_Ordering_using_YANG_Models_Design.md

```diff
@@ -0,0 +1,698 @@
+
+### 3.1.2 Using YANG models to Generate Other Moves
+Check [2.2.3 Generating all possible moves](#223-generating-all-possible-moves) to learn more about what are the possible moves.
+
+YANG models can help with:
+- Lines getting removed which have references, can be substituted by deleting their references first as it is required anyway (check [2.2.3.3 Generating Other Moves](#2233-generating-other-moves) for details). Getting the dependency of some data can be done using [SONiC Libynag wrapper](https://github.com/Azure/sonic-buildimage/blob/master/src/sonic-yang-mgmt/sonic_yang.py), specifically `find_data_dependencies`.
+- Any other moves should be generated using YANG models, if YANG models data are not enough an extension to YANG models can be added to help with generating new moves.
+
+### 3.1.3 Using YANG models to Host Flags for Move Validation
+#### 3.1.3.1 Extension: create-only
+Some fields are not allowed to be modified and can only be created, such as `/PORT/Ethernet0/lanes` which is marked as `CREATE_ONLY` in [opencomputeproject/SAI/blob/master/inc/saiport.h](https://github.com/opencomputeproject/SAI/blob/master/inc/saiport.h#L624):
+```C
+    /**
+     * @brief Hardware Lane list
+     *
+     * @type sai_u32_list_t
+     * @flags MANDATORY_ON_CREATE | CREATE_ONLY | KEY
+     */
+    SAI_PORT_ATTR_HW_LANE_LIST,
+```
+
+The new extension `create-only` will be added to [YANG models SONiC extensions](https://github.com/Azure/sonic-buildimage/blob/master/src/sonic-yang-models/yang-models/sonic-extension.yang):
+```
+extension create-only {
+    description "During apply-patch operation the field can only be created (i.e. added) or re-created (i.e. removed then added) but cannot be modified (i.e. replaced).";
+}
+```
+
+`create-only` field moves will be evaluated as follows:
+1. Fail validation for `replace` move that is updating a field marked with `create-only`.
+2. Fail validation for `replace` move that is updating parent of a field that is marked with `create-only` and the `create-only` field is itself different.
+3. Pass validation for `replace` move that is updating parent of a field that is marked with `create-only` and the `create-only` field is itself the same.
+4. Rule 2,3 above will be also applied to the grand-parent, grand-grand-parent ...etc
+
+
+Let's take the example [A.2 Dynamic Port Breakout 1 to 4](#a2-dynamic-port-breakout-1-to-4) which has the following:
+```json
+[
+    {
+        "op": "replace",
+        "path": "/PORT/Ethernet0/lanes",
+        "value": "65"
+    }
+]
+```
+But the field `lanes` is `create-only` and cannot be modified i.e. replaced. The only way to update the `lanes` field is through deleting then adding its parent the `/PORT/Ethernet0`.
+
+In [port YANG model](https://github.com/Azure/sonic-buildimage/blob/master/src/sonic-yang-models/yang-models/sonic-port.yang), let's mark `lanes` with `create-only`:
+```diff
+ leaf lanes {
+     mandatory true;
++    ext:create-only;
```

Comment:
We need to add processing of this in the sonic-yang-mgmt python library

Reply:
Yes, will have to add this support.